File: app/src/main/java/com/inelasticcollision/recipelink/data/local/dao/RecipeDao.java
package com.inelasticcollision.recipelink.data.local.dao;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import com.inelasticcollision.recipelink.data.local.LocalDataProvider;
import com.inelasticcollision.recipelink.data.models.Recipe;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;

@Dao
public abstract class RecipeDao implements LocalDataProvider {

    @Query("SELECT * FROM recipes ORDER BY title COLLATE NOCASE")
    public abstract io.reactivex.Observable<List<Recipe>> getAllRecipes();

    @Query("SELECT * FROM recipes WHERE favorite = 1 ORDER BY title COLLATE NOCASE")
    public abstract io.reactivex.Observable<List<Recipe>> getFavoriteRecipes();

    public io.reactivex.Observable<List<Recipe>> getRecipesBySearchTerm(@Nullable String searchTerm) {
        if (searchTerm == null || searchTerm.isEmpty()) {
            return getAllRecipes();
        } else {
            return searchRecipes("%" + searchTerm + "%");
        }
    }

    @Query("SELECT * FROM recipes WHERE title LIKE :searchTerm OR tags LIKE :searchTerm ORDER BY title COLLATE NOCASE")
    abstract io.reactivex.Observable<List<Recipe>> searchRecipes(@NonNull String searchTerm);

    @Query("SELECT * FROM recipes WHERE id = :id")
    public abstract Single<Recipe> getRecipeById(@NonNull String id);

    @Insert
    public abstract Completable saveRecipe(@NonNull Recipe recipe);

    @Update
    public abstract Completable updateRecipe(@NonNull Recipe recipe);

    @Query("DELETE from recipes WHERE id = :id")
    public abstract Completable deleteRecipe(@NonNull String id);

}

File: app/src/main/java/com/inelasticcollision/recipelink/data/cache/dao/RecipeDao.kt
package com.inelasticcollision.recipelink.data.cache.dao

import androidx.annotation.VisibleForTesting
import androidx.room.*
import com.inelasticcollision.recipelink.data.model.Recipe
import kotlinx.coroutines.flow.Flow
import java.util.*

@Dao
interface RecipeDao {

    // Recipe methods
    @Query("SELECT * FROM recipes ORDER BY title COLLATE NOCASE ASC")
    fun getAllRecipes(): Flow<List<Recipe>>

    @Query("SELECT * FROM recipes WHERE favorite = 1 ORDER BY title COLLATE NOCASE ASC")
    fun getFavoriteRecipes(): Flow<List<Recipe>>

    @Query("SELECT * FROM recipes WHERE title LIKE :searchTerm OR url LIKE :searchTerm OR tags LIKE :searchTerm ORDER BY title COLLATE NOCASE ASC")
    fun getRecipesBySearchTerm(searchTerm: String): Flow<List<Recipe>>

    @Query("SELECT * FROM recipes WHERE id = :id")
    fun getRecipeById(id: String): Flow<Recipe?>

    @Query("UPDATE recipes SET favorite = ((favorite | 1) - (favorite & 1)), last_modified = :lastModified WHERE id = :id")
    suspend fun toggleFavoriteRecipeWithId(id: String, lastModified: Long = Date().time)

    @Insert(onConflict = OnConflictStrategy.ABORT)
    suspend fun insertRecipe(recipe: Recipe)

    @Update(onConflict = OnConflictStrategy.ABORT)
    suspend fun updateRecipe(recipe: Recipe)

    @Query("DELETE FROM recipes WHERE id = :id")
    suspend fun deleteRecipeById(id: String)

    @VisibleForTesting
    @Insert(onConflict = OnConflictStrategy.ABORT)
    suspend fun insertRecipes(recipes: List<Recipe>)
}